BeatSaberTheater/Services/AuthTokenService.cs

```diff
@@ -0,0 +1,139 @@
+
+    private readonly LoggingService _loggingService;
+    private IPlatformUserModel? _platformUserModel;
+    private CancellationTokenSource? _refreshTokenCts;
+    private DateTime _tokenExpiryUtc = DateTime.MinValue;
+
+    public AuthTokenService(LoggingService loggingService)
+    {
+        _loggingService = loggingService;
+    }
+
+    public void Initialize()
+    {
+        BSEvents.lateMenuSceneLoadedFresh += OnMenuSceneLoadedFresh;
+    }
+
+    private void OnMenuSceneLoadedFresh(ScenesTransitionSetupDataSO? scenesTransition)
+    {
+        Task.Run(RefreshAuthToken);
+    }
+
+    private async Task RefreshAuthToken()
+    {
+        _loggingService.Debug("Attempting rotating auth token");
+
+        if (_platformUserModel is null)
+        {
+            var platformUserModel = Resources
+                .FindObjectsOfTypeAll<PlatformLeaderboardsModel>()
+                .Select(l => l._platformUserModel)
+                .Last(x => x != null);
+            _platformUserModel = platformUserModel;
+        }
+
+        var platformUserToken = await _platformUserModel.GetUserAuthToken();
+        _loggingService.Debug("PLATFORM USER: " + platformUserToken.validPlatformEnvironment);
+        _loggingService.Debug("PLATFORM USER TOKEN: " + platformUserToken.token);
+        var token = platformUserToken.token;
+
+        var authTokenResponse = await RetrieveUserToken(token);
+        CurrentAuthToken = authTokenResponse.Token;
+
+        // Parse JWT expiration (exp) claim if present
+        _tokenExpiryUtc = ParseJwtExpiry(CurrentAuthToken) ?? DateTime.UtcNow.AddMinutes(55); // Fallback to 55 minutes if expiry could not be parsed from JWT
+
+        _loggingService.Debug($"Current auth token has been rotated: {CurrentAuthToken}");
+        _loggingService.Debug($"Token expires at {_tokenExpiryUtc:u}");
+
+        ScheduleNextRefresh();
+    }
+
+    private void ScheduleNextRefresh()
+    {
+        _refreshTokenCts?.Cancel();
+        _refreshTokenCts = new CancellationTokenSource();
+
+        var timeUntilExpiry = _tokenExpiryUtc - DateTime.UtcNow;
+        var refreshDelay = timeUntilExpiry - TimeSpan.FromMinutes(5); // refresh 5 minutes before expiry
+        if (refreshDelay < TimeSpan.Zero)
+            refreshDelay = TimeSpan.FromMinutes(1); // refresh soon if token is near-expired
+
+        _ = Task.Run(async () =>
+        {
+            try
+            {
+                _loggingService.Debug($"Scheduling next auth token refresh in {refreshDelay.TotalMinutes:F1} minutes");
+                await Task.Delay(refreshDelay, _refreshTokenCts.Token);
+                await RefreshAuthToken();
+            }
+            catch (TaskCanceledException)
+            {
+                // expected when Dispose() or new refresh
+            }
+            catch (Exception ex)
+            {
+                _loggingService.Error($"Error scheduling next token refresh: {ex}");
+            }
+        }, _refreshTokenCts.Token);
```

Comment:
Would creating one thread or task on initialization of this service, with a while loop (until cancellation) requested not be way easier and less error prone? It could contain a while loop which checks:

- If token will be cancelled soon
  - Then do refresh
- Sleep 10-60 seconds

The current implementation spawns a task which will be dormant for a while, which then calls a method that spawns another task that does the same thing.

Just an idea.